ambry-router/src/main/java/com/github/ambry/router/CompressionService.java

```diff
@@ -250,32 +248,14 @@ public ByteBuf compressChunk(ByteBuf chunkBuffer, boolean isFullChunk) {
       return null;
     }
 
-    // Convert ByteBuf to byte[] before applying compression.
-    byte[] sourceBuffer;
-    int sourceBufferOffset;
-    try {
-      Triple<byte[], Integer, Integer> bufferInfo = convertByteBufToByteArray(chunkBuffer);
-      sourceBuffer = bufferInfo.getLeft();
-      sourceBufferOffset = bufferInfo.getMiddle();
-      sourceDataSize = bufferInfo.getRight();
-    } catch (Exception ex) {
-      logger.error(String.format("Cannot compress source data because failed to convert ByteBuf to byte[]."
-              + "chunkBuffer.hasArray() = %s, chunkBuffer.readableBytes = %d, chunkBuffer.arrayOffset = %d",
-          chunkBuffer.hasArray(), chunkBuffer.readableBytes(), chunkBuffer.arrayOffset()), ex);
-
-      // Emit metrics for count how many compressions skipped due to buffer conversion error.
-      compressionMetrics.compressErrorRate.mark();
-      compressionMetrics.compressErrorBufferConversion.inc();
-      return null;
-    }
-
     // Apply compression.
     CompressionMetrics.AlgorithmMetrics algorithmMetrics = compressionMetrics.getAlgorithmMetrics(defaultCompressor.getAlgorithmName());
-    Pair<Integer, byte[]> compressResult;
+    ByteBuf newChunkBuffer = combineBuffer(chunkBuffer);
+    ByteBuffer compressedBuffer;
     try {
       algorithmMetrics.compressRate.mark();
       long startTime = System.nanoTime();
-      compressResult = defaultCompressor.compress(sourceBuffer, sourceBufferOffset, sourceDataSize);
+      compressedBuffer = defaultCompressor.compress(newChunkBuffer.nioBuffer(), outputDirectBuffer);
```

Comment:
like we discussed, we should create ByteBuf for destination before calling this method.